tests/baselines/reference/privateNameNotAllowedOutsideClass.errors.txt

```diff
@@ -0,0 +1,15 @@
+tests/cases/conformance/classes/members/privateNames/privateNameNotAllowedOutsideClass.ts(3,7): error TS1134: Variable declaration expected.
+tests/cases/conformance/classes/members/privateNames/privateNameNotAllowedOutsideClass.ts(3,12): error TS1134: Variable declaration expected.
+tests/cases/conformance/classes/members/privateNames/privateNameNotAllowedOutsideClass.ts(3,14): error TS1134: Variable declaration expected.
+
+
+==== tests/cases/conformance/classes/members/privateNames/privateNameNotAllowedOutsideClass.ts (3 errors) ====
+    // @target es6
+    
+    const #foo = 3;
+          ~~~~
+!!! error TS1134: Variable declaration expected.

```

Comment:
a specific, non-parse error message would be better here. It would be worthwhile parsing privates in this location to make that happen.

Reply:
Is the suggestion to:
- modify the parser so that private names are allowed in all the same places as identifiers
- then, in the checker, make diagnostic errors whenever there's a private name where there shouldn't be

For the checker implementation, on @rbuckton 's guidance, I tried to treat private names similarly to computed properties. These are the errors when using a computed property-like syntax where an identifier would be expected:

```ts
const ["a"] = 3
                   ^ Error: declaration or statement expected
            ^ Error: Array element destructuring pattern expected
```

Reply:
Agreed. As a user I would appreciate an error message that says something about how private names are only allowed in class bodies.

Which approach would you recommend for fixing it? I jotted some ideas down here: https://github.com/microsoft/TypeScript/pull/30829#discussion_r299706018

Reply:
@sandersn @rbuckton @DanielRosenwasser hoping we can get consensus on the approach in the checker. I summarized the approaches and trade-offs to the best of my knowledge here: https://github.com/microsoft/TypeScript/pull/30829

Reply:
whoops: correct link to the summary: https://github.com/microsoft/TypeScript/pull/30829#discussion_r299706018

Reply:
I'll let you know when we've discussed it and decided on something.

Reply:
Thanks Nathan and Ron! Your consolidated feedback on the checker approach is [here](https://github.com/microsoft/TypeScript/pull/30829#issuecomment-509839268). I'll consolidate answers here since that comment doesn't allow threading.

re consolidated error reporting with that for `private` modifier, particularly for `propertiesRelatedTo`:
- I'll review the other error handling and look for points of consolidation.
- Reply re `propertiesRelatedTo`: https://github.com/microsoft/TypeScript/pull/30829/files#r302071097. I think the concerns there re consolidation might generalize.

re index-access code: If the part we're talking about changing is `getLiteralTypeForPropertyName`:
- it's true that the `return neverType` isn't needed for banning `foo[#bar]` syntax
- However, the `return neverType` seems to be load-bearing with regard to the the behavior of `keyof` with private names:
    - If we remove the `if` entirely, then `checker.ts` doesn't type-check
    - If I add an `assert`, the `assert` gets hit when the `privateNamesAndkeyof` test case is run.

- re tests: thanks Ron! will incorporate those test cases

Reply:
The check should probably go in getLiteralTypeFromProperty, where normal privates are excluded:

```ts
        function getLiteralTypeFromProperty(prop: Symbol, include: TypeFlags) {
            if (!(getDeclarationModifierFlagsFromSymbol(prop) & ModifierFlags.NonPublicAccessibilityModifier)) {
```

Does that work? You probably still need an assert, or a cast in getLiteralTypeFromProperty, to make the compiler happy.

Reply:
discussed elsewhere, resolving